### DISCUSSION TITLE
Help with first graphql.net code

### DISCUSSION THREAD

**Discussion by PhilPJL** [graphql-dotnet/graphql-client#611] [Q&A] [answered]
Hi,

I'm trying to call Octopus Energy's GraphQL api using LinqPad, as so:

```
async void Main()
{
	var options = new GraphQLHttpClientOptions
	{
		EndPoint = new Uri("https://api.octopus.energy/v1/graphql"),
// adding delegating handler so I can see what's going on
		HttpMessageHandler = new MyHandler{ InnerHandler = new HttpClientHandler() },
	};
	
	var client = new GraphQLHttpClient(options, new NewtonsoftJsonSerializer());

	client.HttpClient.DefaultRequestHeaders.Authorization = new AuthenticationHeaderValue("Basic", "<api key here>");

	var request = new GraphQLHttpRequest(
			@"query viewer {
				viewer {
					fullName
					accounts {
						number	
					}
				}
			}",
			null,
			"viewer");

	var response = await client.SendQueryAsync<Response>(request);

	response.Dump();
}

public class Response
{
	public string FullName { get; set; }
}

public class MyHandler : DelegatingHandler
{
	protected override HttpResponseMessage Send(HttpRequestMessage request, CancellationToken cancellationToken)
	{
		request.Dump();
		
		return base.Send(request, cancellationToken);
	}

	protected async override Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
	{
		request.Dump();

		return await base.SendAsync(request, cancellationToken);
	}
}
```

the result I get is "Must provide query string." which is correct given the generated request Uri

![image](https://github.com/graphql-dotnet/graphql-dotnet/assets/271656/acf43f17-482f-416e-915d-f6dc96db7b4c)

I must be doing something obviously wrong.  Any ideas?

**Comment by Shane32** [comment]
Which GraphQL nuget package(s) and version(s) are you using ?

**Reply by PhilPJL** [reply]
![image](https://github.com/graphql-dotnet/graphql-dotnet/assets/271656/0e56a18c-8e51-4d4c-a8d1-b918d4395a93)

**Comment by PhilPJL** [comment]
@Shane32 Any ideas?  Could you try that code in LinqPad.  It won't work without an api key but you should at least get a different error if it's working.

**Comment by Shane32** [comment]
Well, I support the GraphQL and GraphQL.Server repos, not the GraphQL.Client repo.  Nor do I use this package in any of my own projects.  And LinqPad 8 free edition doesn't allow me to import NuGet packages.

I don't see anything obviously wrong here.  Maybe if you want to refactor into a console app, I could take a look.  It should not be hard to diagnose.  Octopus Energy themselves provide a GraphiQL interface here: https://api.octopus.energy/v1/graphql

Perhaps @rose-a can give you a few pointers.

**Reply by PhilPJL** [reply]
https://api.octopus.energy/v1/graphql works fine, but I wanted to do something in C#.  I'll try the console app suggestion.

**Comment by PhilPJL** [comment]
I tried again with a console app.  Same result.  @rose-a any idea what I'm doing wrong?

**Reply by Shane32** [reply]
Want to post the console app?

**Comment by PhilPJL** [comment]
Ok I've got further by 
1) changing GraphQLHttpRequest to GraphQLRequest as per your example client app.  This didn't help though.
2) add a trailing / to the URL.  So "https://api.octopus.energy/v1/graphql/".  This now gives me an authorization error, so that's progress.  Seems like a bug that it doesn't handle a URL without a trailing /.

I also note that your example client app fails with an exception 
"The requested name is valid, but no data of the requested type was found. (swapi.apis.guru:443)"

**Comment by PhilPJL** [comment]
The solution in the end was to add a trailing / to the URL

**Reply by Shane32** [reply, marked as answer]
Ok well that’s due to bad documentation on their part then. Technically the two urls are different - the server can implement different responses whether or not there’s a slash at the end. Most web servers automatically compensate. Here they host GraphiQL at the url without a slash and the GraphQL server at the address with a slash it would seem. At least that’s my best guess without personally testing it.

**Reply by PhilPJL** [reply]
Hmmm, bad design on Octopus's part then I think.
